File: tests/integration/interop_test.go
// Package integration provides cross-runtime interoperability tests.
//
// These tests verify that Go, TypeScript, and Rust runtimes produce
// identical binary encodings and can decode each other's output.
package integration

import (
	"bytes"
	"encoding/hex"
	"math"
	"os"
	"path/filepath"
	"testing"

	"github.com/cramberry/cramberry-go/pkg/cramberry"
	"github.com/cramberry/cramberry-go/tests/integration/gen"
)

// TestData contains all the test cases used for cross-runtime verification.
var TestData = struct {
	ScalarTypes     *interop.ScalarTypes
	RepeatedTypes   *interop.RepeatedTypes
	NestedMessage   *interop.NestedMessage
	ComplexTypes    *interop.ComplexTypes
	EdgeCases       *interop.EdgeCases
	AllFieldNumbers *interop.AllFieldNumbers
}{
	ScalarTypes: &interop.ScalarTypes{
		BoolVal:    true,
		Int32Val:   -42,
		Int64Val:   -9223372036854775807,
		Uint32Val:  4294967295,
		Uint64Val:  18446744073709551615,
		Float32Val: 3.14159,
		Float64Val: 2.718281828459045,
		StringVal:  "hello, cramberry!",
		BytesVal:   []byte{0xde, 0xad, 0xbe, 0xef},
	},
	RepeatedTypes: &interop.RepeatedTypes{
		Int32List:  []int32{1, -2, 3, -4, 5},
		StringList: []string{"alpha", "beta", "gamma"},
		BytesList:  [][]byte{{0x01, 0x02}, {0x03, 0x04, 0x05}},
	},
	NestedMessage: &interop.NestedMessage{
		Name:  "nested",
		Value: 123,
	},
	ComplexTypes: &interop.ComplexTypes{
		Status: interop.StatusActive,
		OptionalNested: &interop.NestedMessage{
			Name:  "optional",
			Value: 456,
		},
		RequiredNested: interop.NestedMessage{
			Name:  "required",
			Value: 789,
		},
		NestedList: []interop.NestedMessage{
			{Name: "first", Value: 1},
			{Name: "second", Value: 2},
		},
		StringIntMap: map[string]int32{
			"one":   1,
			"two":   2,
			"three": 3,
		},
		IntStringMap: map[int32]string{
			1: "one",
			2: "two",
			3: "three",
		},
	},
	EdgeCases: &interop.EdgeCases{
		ZeroInt:       0,
		NegativeOne:   -1,
		MaxInt32:      math.MaxInt32,
		MinInt32:      math.MinInt32,
		MaxInt64:      math.MaxInt64,
		MinInt64:      math.MinInt64,
		MaxUint32:     math.MaxUint32,
		MaxUint64:     math.MaxUint64,
		EmptyString:   "",
		UnicodeString: "Hello, 世界! 🎉",
		EmptyBytes:    []byte{},
	},
	AllFieldNumbers: &interop.AllFieldNumbers{
		Field1:    100,
		Field15:   1500,
		Field16:   1600,
		Field127:  12700,
		Field128:  12800,
		Field1000: 100000,
	},
}

const goldenDir = "../golden"

// TestScalarTypesEncodeDecode tests encoding and decoding of scalar types.
func TestScalarTypesEncodeDecode(t *testing.T) {
	data, err := cramberry.Marshal(TestData.ScalarTypes)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	t.Logf("ScalarTypes encoded size: %d bytes", len(data))
	t.Logf("ScalarTypes hex: %s", hex.EncodeToString(data))

	var decoded interop.ScalarTypes
	if err := cramberry.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	// Verify all fields
	if decoded.BoolVal != TestData.ScalarTypes.BoolVal {
		t.Errorf("BoolVal mismatch: got %v, want %v", decoded.BoolVal, TestData.ScalarTypes.BoolVal)
	}
	if decoded.Int32Val != TestData.ScalarTypes.Int32Val {
		t.Errorf("Int32Val mismatch: got %v, want %v", decoded.Int32Val, TestData.ScalarTypes.Int32Val)
	}
	if decoded.Int64Val != TestData.ScalarTypes.Int64Val {
		t.Errorf("Int64Val mismatch: got %v, want %v", decoded.Int64Val, TestData.ScalarTypes.Int64Val)
	}
	if decoded.Uint32Val != TestData.ScalarTypes.Uint32Val {
		t.Errorf("Uint32Val mismatch: got %v, want %v", decoded.Uint32Val, TestData.ScalarTypes.Uint32Val)
	}
	if decoded.Uint64Val != TestData.ScalarTypes.Uint64Val {
		t.Errorf("Uint64Val mismatch: got %v, want %v", decoded.Uint64Val, TestData.ScalarTypes.Uint64Val)
	}
	if decoded.Float32Val != TestData.ScalarTypes.Float32Val {
		t.Errorf("Float32Val mismatch: got %v, want %v", decoded.Float32Val, TestData.ScalarTypes.Float32Val)
	}
	if decoded.Float64Val != TestData.ScalarTypes.Float64Val {
		t.Errorf("Float64Val mismatch: got %v, want %v", decoded.Float64Val, TestData.ScalarTypes.Float64Val)
	}
	if decoded.StringVal != TestData.ScalarTypes.StringVal {
		t.Errorf("StringVal mismatch: got %v, want %v", decoded.StringVal, TestData.ScalarTypes.StringVal)
	}
	if !bytes.Equal(decoded.BytesVal, TestData.ScalarTypes.BytesVal) {
		t.Errorf("BytesVal mismatch: got %v, want %v", decoded.BytesVal, TestData.ScalarTypes.BytesVal)
	}
}

// TestRepeatedTypesEncodeDecode tests encoding and decoding of repeated types.
func TestRepeatedTypesEncodeDecode(t *testing.T) {
	data, err := cramberry.Marshal(TestData.RepeatedTypes)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	t.Logf("RepeatedTypes encoded size: %d bytes", len(data))
	t.Logf("RepeatedTypes hex: %s", hex.EncodeToString(data))

	var decoded interop.RepeatedTypes
	if err := cramberry.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	// Verify int32 list
	if len(decoded.Int32List) != len(TestData.RepeatedTypes.Int32List) {
		t.Errorf("Int32List length mismatch: got %d, want %d", len(decoded.Int32List), len(TestData.RepeatedTypes.Int32List))
	}
	for i, v := range TestData.RepeatedTypes.Int32List {
		if decoded.Int32List[i] != v {
			t.Errorf("Int32List[%d] mismatch: got %d, want %d", i, decoded.Int32List[i], v)
		}
	}

	// Verify string list
	if len(decoded.StringList) != len(TestData.RepeatedTypes.StringList) {
		t.Errorf("StringList length mismatch: got %d, want %d", len(decoded.StringList), len(TestData.RepeatedTypes.StringList))
	}
	for i, v := range TestData.RepeatedTypes.StringList {
		if decoded.StringList[i] != v {
			t.Errorf("StringList[%d] mismatch: got %q, want %q", i, decoded.StringList[i], v)
		}
	}
}

// TestNestedMessageEncodeDecode tests encoding and decoding of nested messages.
func TestNestedMessageEncodeDecode(t *testing.T) {
	data, err := cramberry.Marshal(TestData.NestedMessage)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	t.Logf("NestedMessage encoded size: %d bytes", len(data))
	t.Logf("NestedMessage hex: %s", hex.EncodeToString(data))

	var decoded interop.NestedMessage
	if err := cramberry.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	if decoded.Name != TestData.NestedMessage.Name {
		t.Errorf("Name mismatch: got %q, want %q", decoded.Name, TestData.NestedMessage.Name)
	}
	if decoded.Value != TestData.NestedMessage.Value {
		t.Errorf("Value mismatch: got %d, want %d", decoded.Value, TestData.NestedMessage.Value)
	}
}

// TestComplexTypesEncodeDecode tests encoding and decoding of complex types.
func TestComplexTypesEncodeDecode(t *testing.T) {
	data, err := cramberry.Marshal(TestData.ComplexTypes)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	t.Logf("ComplexTypes encoded size: %d bytes", len(data))
	t.Logf("ComplexTypes hex: %s", hex.EncodeToString(data))

	var decoded interop.ComplexTypes
	if err := cramberry.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	if decoded.Status != TestData.ComplexTypes.Status {
		t.Errorf("Status mismatch: got %v, want %v", decoded.Status, TestData.ComplexTypes.Status)
	}

	if decoded.OptionalNested == nil {
		t.Error("OptionalNested is nil, expected value")
	} else {
		if decoded.OptionalNested.Name != TestData.ComplexTypes.OptionalNested.Name {
			t.Errorf("OptionalNested.Name mismatch")
		}
	}

	if decoded.RequiredNested.Name != TestData.ComplexTypes.RequiredNested.Name {
		t.Errorf("RequiredNested.Name mismatch")
	}

	if len(decoded.NestedList) != len(TestData.ComplexTypes.NestedList) {
		t.Errorf("NestedList length mismatch")
	}

	// Verify maps (note: map ordering may differ, but values should match)
	if len(decoded.StringIntMap) != len(TestData.ComplexTypes.StringIntMap) {
		t.Errorf("StringIntMap length mismatch")
	}
	for k, v := range TestData.ComplexTypes.StringIntMap {
		if decoded.StringIntMap[k] != v {
			t.Errorf("StringIntMap[%q] mismatch: got %d, want %d", k, decoded.StringIntMap[k], v)
		}
	}
}

// TestEdgeCasesEncodeDecode tests encoding and decoding of edge case values.
func TestEdgeCasesEncodeDecode(t *testing.T) {
	data, err := cramberry.Marshal(TestData.EdgeCases)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	t.Logf("EdgeCases encoded size: %d bytes", len(data))
	t.Logf("EdgeCases hex: %s", hex.EncodeToString(data))

	var decoded interop.EdgeCases
	if err := cramberry.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	if decoded.ZeroInt != 0 {
		t.Errorf("ZeroInt mismatch: got %d, want 0", decoded.ZeroInt)
	}
	if decoded.NegativeOne != -1 {
		t.Errorf("NegativeOne mismatch: got %d, want -1", decoded.NegativeOne)
	}
	if decoded.MaxInt32 != math.MaxInt32 {
		t.Errorf("MaxInt32 mismatch: got %d, want %d", decoded.MaxInt32, math.MaxInt32)
	}
	if decoded.MinInt32 != math.MinInt32 {
		t.Errorf("MinInt32 mismatch: got %d, want %d", decoded.MinInt32, math.MinInt32)
	}
	if decoded.MaxInt64 != math.MaxInt64 {
		t.Errorf("MaxInt64 mismatch: got %d, want %d", decoded.MaxInt64, math.MaxInt64)
	}
	if decoded.MinInt64 != math.MinInt64 {
		t.Errorf("MinInt64 mismatch: got %d, want %d", decoded.MinInt64, math.MinInt64)
	}
	if decoded.MaxUint32 != math.MaxUint32 {
		t.Errorf("MaxUint32 mismatch")
	}
	if decoded.MaxUint64 != math.MaxUint64 {
		t.Errorf("MaxUint64 mismatch")
	}
	if decoded.UnicodeString != TestData.EdgeCases.UnicodeString {
		t.Errorf("UnicodeString mismatch: got %q, want %q", decoded.UnicodeString, TestData.EdgeCases.UnicodeString)
	}
}

// TestAllFieldNumbersEncodeDecode tests encoding and decoding with various field numbers.
func TestAllFieldNumbersEncodeDecode(t *testing.T) {
	data, err := cramberry.Marshal(TestData.AllFieldNumbers)
	if err != nil {
		t.Fatalf("Marshal failed: %v", err)
	}

	t.Logf("AllFieldNumbers encoded size: %d bytes", len(data))
	t.Logf("AllFieldNumbers hex: %s", hex.EncodeToString(data))

	var decoded interop.AllFieldNumbers
	if err := cramberry.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("Unmarshal failed: %v", err)
	}

	if decoded.Field1 != TestData.AllFieldNumbers.Field1 {
		t.Errorf("Field1 mismatch")
	}
	if decoded.Field15 != TestData.AllFieldNumbers.Field15 {
		t.Errorf("Field15 mismatch")
	}
	if decoded.Field16 != TestData.AllFieldNumbers.Field16 {
		t.Errorf("Field16 mismatch")
	}
	if decoded.Field127 != TestData.AllFieldNumbers.Field127 {
		t.Errorf("Field127 mismatch")
	}
	if decoded.Field128 != TestData.AllFieldNumbers.Field128 {
		t.Errorf("Field128 mismatch")
	}
	if decoded.Field1000 != TestData.AllFieldNumbers.Field1000 {
		t.Errorf("Field1000 mismatch")
	}
}

// TestGenerateGoldenFiles generates golden byte files for cross-runtime testing.
// Run with: go test -v -run TestGenerateGoldenFiles -generate-golden
func TestGenerateGoldenFiles(t *testing.T) {
	if os.Getenv("GENERATE_GOLDEN") != "1" {
		t.Skip("Set GENERATE_GOLDEN=1 to regenerate golden files")
	}

	if err := os.MkdirAll(goldenDir, 0755); err != nil {
		t.Fatalf("Failed to create golden dir: %v", err)
	}

	testCases := []struct {
		name string
		data interface{}
	}{
		{"scalar_types", TestData.ScalarTypes},
		{"repeated_types", TestData.RepeatedTypes},
		{"nested_message", TestData.NestedMessage},
		{"complex_types", TestData.ComplexTypes},
		{"edge_cases", TestData.EdgeCases},
		{"all_field_numbers", TestData.AllFieldNumbers},
	}

	for _, tc := range testCases {
		data, err := cramberry.Marshal(tc.data)
		if err != nil {
			t.Errorf("Failed to marshal %s: %v", tc.name, err)
			continue
		}

		path := filepath.Join(goldenDir, tc.name+".bin")
		if err := os.WriteFile(path, data, 0644); err != nil {
			t.Errorf("Failed to write %s: %v", path, err)
			continue
		}

		// Also write hex for easier inspection
		hexPath := filepath.Join(goldenDir, tc.name+".hex")
		if err := os.WriteFile(hexPath, []byte(hex.EncodeToString(data)), 0644); err != nil {
			t.Errorf("Failed to write %s: %v", hexPath, err)
		}

		t.Logf("Generated %s (%d bytes)", path, len(data))
	}
}

// TestVerifyGoldenFiles verifies that current encoding matches golden files.
func TestVerifyGoldenFiles(t *testing.T) {
	testCases := []struct {
		name string
		data interface{}
	}{
		{"scalar_types", TestData.ScalarTypes},
		{"repeated_types", TestData.RepeatedTypes},
		{"nested_message", TestData.NestedMessage},
		{"complex_types", TestData.ComplexTypes},
		{"edge_cases", TestData.EdgeCases},
		{"all_field_numbers", TestData.AllFieldNumbers},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			path := filepath.Join(goldenDir, tc.name+".bin")
			golden, err := os.ReadFile(path)
			if os.IsNotExist(err) {
				t.Skipf("Golden file not found: %s (run with GENERATE_GOLDEN=1 to create)", path)
				return
			}
			if err != nil {
				t.Fatalf("Failed to read golden file: %v", err)
			}

			encoded, err := cramberry.Marshal(tc.data)
			if err != nil {
				t.Fatalf("Failed to marshal: %v", err)
			}

			if !bytes.Equal(encoded, golden) {
				t.Errorf("Encoding mismatch for %s\nGot:  %s\nWant: %s",
					tc.name, hex.EncodeToString(encoded), hex.EncodeToString(golden))
			}
		})
	}
}

File: test/integration/ts/interop.test.ts
/**
 * Cross-runtime interoperability tests for TypeScript.
 *
 * These tests verify that TypeScript runtime produces identical binary
 * encodings to Go and can decode Go-generated golden files.
 */

import { describe, it, expect, beforeAll } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';
import { Writer, Reader, WireType } from '../../../typescript/src';
import {
  Status,
  ScalarTypes,
  RepeatedTypes,
  NestedMessage,
  ComplexTypes,
  EdgeCases,
  AllFieldNumbers,
} from './interop';

const GOLDEN_DIR = path.join(__dirname, '..', '..', 'golden');

// Test data matching Go's TestData
const TestData = {
  scalarTypes: {
    boolVal: true,
    int32Val: -42,
    int64Val: -9223372036854775807n,
    uint32Val: 4294967295,
    uint64Val: 18446744073709551615n,
    float32Val: 3.14159,
    float64Val: 2.718281828459045,
    stringVal: 'hello, cramberry!',
    bytesVal: new Uint8Array([0xde, 0xad, 0xbe, 0xef]),
  } as ScalarTypes,

  repeatedTypes: {
    int32List: [1, -2, 3, -4, 5],
    stringList: ['alpha', 'beta', 'gamma'],
    bytesList: [new Uint8Array([0x01, 0x02]), new Uint8Array([0x03, 0x04, 0x05])],
  } as RepeatedTypes,

  nestedMessage: {
    name: 'nested',
    value: 123,
  } as NestedMessage,

  complexTypes: {
    status: Status.Active,
    optionalNested: { name: 'optional', value: 456 },
    requiredNested: { name: 'required', value: 789 },
    nestedList: [
      { name: 'first', value: 1 },
      { name: 'second', value: 2 },
    ],
    stringIntMap: { one: 1, two: 2, three: 3 },
    intStringMap: new Map([[1, 'one'], [2, 'two'], [3, 'three']]),
  } as ComplexTypes,

  edgeCases: {
    zeroInt: 0,
    negativeOne: -1,
    maxInt32: 2147483647,
    minInt32: -2147483648,
    maxInt64: 9223372036854775807n,
    minInt64: -9223372036854775808n,
    maxUint32: 4294967295,
    maxUint64: 18446744073709551615n,
    emptyString: '',
    unicodeString: 'Hello, 世界! 🎉',
    emptyBytes: new Uint8Array([]),
  } as EdgeCases,

  allFieldNumbers: {
    field1: 100,
    field15: 1500,
    field16: 1600,
    field127: 12700,
    field128: 12800,
    field1000: 100000,
  } as AllFieldNumbers,
};

// Encoder functions for each message type using V2 format
// V2 uses end markers instead of field counts, and SVarint wire type for signed ints
function encodeNestedMessage(writer: Writer, msg: NestedMessage): void {
  // Field 1: name (string)
  writer.writeTag(1, WireType.Bytes);
  writer.writeString(msg.name);

  // Field 2: value (int32) - uses SVarint wire type
  writer.writeTag(2, WireType.SVarint);
  writer.writeSVarint(msg.value);

  // End marker
  writer.writeEndMarker();
}

function encodeScalarTypes(writer: Writer, msg: ScalarTypes): void {
  // V2 format: no field count prefix, uses end marker

  // Field 1: boolVal
  writer.writeTag(1, WireType.Varint);
  writer.writeBool(msg.boolVal);

  // Field 2: int32Val - uses SVarint wire type
  writer.writeTag(2, WireType.SVarint);
  writer.writeSVarint(msg.int32Val);

  // Field 3: int64Val - uses SVarint wire type
  writer.writeTag(3, WireType.SVarint);
  writer.writeSVarint64(msg.int64Val);

  // Field 4: uint32Val
  writer.writeTag(4, WireType.Varint);
  writer.writeVarint(msg.uint32Val);

  // Field 5: uint64Val
  writer.writeTag(5, WireType.Varint);
  writer.writeVarint64(msg.uint64Val);

  // Field 6: float32Val
  writer.writeTag(6, WireType.Fixed32);
  writer.writeFloat32(msg.float32Val);

  // Field 7: float64Val
  writer.writeTag(7, WireType.Fixed64);
  writer.writeFloat64(msg.float64Val);

  // Field 8: stringVal
  writer.writeTag(8, WireType.Bytes);
  writer.writeString(msg.stringVal);

  // Field 9: bytesVal
  writer.writeTag(9, WireType.Bytes);
  writer.writeLengthPrefixedBytes(msg.bytesVal);

  // End marker
  writer.writeEndMarker();
}

function encodeRepeatedTypes(writer: Writer, msg: RepeatedTypes): void {
  // V2 format: no field count prefix, uses end marker

  // Field 1: int32List
  writer.writeTag(1, WireType.Bytes);
  const int32Writer = new Writer();
  int32Writer.writeVarint(msg.int32List.length);
  for (const v of msg.int32List) {
    int32Writer.writeSVarint(v);
  }
  writer.writeLengthPrefixedBytes(int32Writer.bytes());

  // Field 2: stringList
  writer.writeTag(2, WireType.Bytes);
  const strWriter = new Writer();
  strWriter.writeVarint(msg.stringList.length);
  for (const v of msg.stringList) {
    strWriter.writeString(v);
  }
  writer.writeLengthPrefixedBytes(strWriter.bytes());

  // Field 3: bytesList
  writer.writeTag(3, WireType.Bytes);
  const bytesWriter = new Writer();
  bytesWriter.writeVarint(msg.bytesList.length);
  for (const v of msg.bytesList) {
    bytesWriter.writeLengthPrefixedBytes(v);
  }
  writer.writeLengthPrefixedBytes(bytesWriter.bytes());

  // End marker
  writer.writeEndMarker();
}

function encodeEdgeCases(writer: Writer, msg: EdgeCases): void {
  // V2 format: no field count prefix, uses end marker
  // Only encode non-zero/non-empty fields to match Go's omitempty behavior

  // Field 2: negativeOne - uses SVarint wire type
  writer.writeTag(2, WireType.SVarint);
  writer.writeSVarint(msg.negativeOne);

  // Field 3: maxInt32 - uses SVarint wire type
  writer.writeTag(3, WireType.SVarint);
  writer.writeSVarint(msg.maxInt32);

  // Field 4: minInt32 - uses SVarint wire type
  writer.writeTag(4, WireType.SVarint);
  writer.writeSVarint(msg.minInt32);

  // Field 5: maxInt64 - uses SVarint wire type
  writer.writeTag(5, WireType.SVarint);
  writer.writeSVarint64(msg.maxInt64);

  // Field 6: minInt64 - uses SVarint wire type
  writer.writeTag(6, WireType.SVarint);
  writer.writeSVarint64(msg.minInt64);

  // Field 7: maxUint32
  writer.writeTag(7, WireType.Varint);
  writer.writeVarint(msg.maxUint32);

  // Field 8: maxUint64
  writer.writeTag(8, WireType.Varint);
  writer.writeVarint64(msg.maxUint64);

  // Field 10: unicodeString
  writer.writeTag(10, WireType.Bytes);
  writer.writeString(msg.unicodeString);

  // End marker
  writer.writeEndMarker();
}

function encodeAllFieldNumbers(writer: Writer, msg: AllFieldNumbers): void {
  // V2 format: no field count prefix, uses end marker
  // All int32 fields use SVarint wire type

  writer.writeTag(1, WireType.SVarint);
  writer.writeSVarint(msg.field1);

  writer.writeTag(15, WireType.SVarint);
  writer.writeSVarint(msg.field15);

  writer.writeTag(16, WireType.SVarint);
  writer.writeSVarint(msg.field16);

  writer.writeTag(127, WireType.SVarint);
  writer.writeSVarint(msg.field127);

  writer.writeTag(128, WireType.SVarint);
  writer.writeSVarint(msg.field128);

  writer.writeTag(1000, WireType.SVarint);
  writer.writeSVarint(msg.field1000);

  // End marker
  writer.writeEndMarker();
}

// Decoder functions - V2 format uses end marker (fieldNumber === 0) instead of field count
function decodeNestedMessage(reader: Reader): NestedMessage {
  const result: Partial<NestedMessage> = {};

  while (reader.hasMore) {
    const tag = reader.readTag();
    if (tag.fieldNumber === 0) break; // End marker

    switch (tag.fieldNumber) {
      case 1:
        result.name = reader.readString();
        break;
      case 2:
        result.value = reader.readSVarint();
        break;
      default:
        reader.skipField(tag.wireType);
    }
  }

  return result as NestedMessage;
}

function decodeScalarTypes(reader: Reader): ScalarTypes {
  const result: Partial<ScalarTypes> = {};

  while (reader.hasMore) {
    const tag = reader.readTag();
    if (tag.fieldNumber === 0) break; // End marker

    switch (tag.fieldNumber) {
      case 1:
        result.boolVal = reader.readBool();
        break;
      case 2:
        result.int32Val = reader.readSVarint();
        break;
      case 3:
        result.int64Val = reader.readSVarint64();
        break;
      case 4:
        result.uint32Val = reader.readVarint();
        break;
      case 5:
        result.uint64Val = reader.readVarint64();
        break;
      case 6:
        result.float32Val = reader.readFloat32();
        break;
      case 7:
        result.float64Val = reader.readFloat64();
        break;
      case 8:
        result.stringVal = reader.readString();
        break;
      case 9:
        result.bytesVal = reader.readLengthPrefixedBytes();
        break;
      default:
        reader.skipField(tag.wireType);
    }
  }

  return result as ScalarTypes;
}

// Helper to convert Uint8Array to hex string
function toHex(data: Uint8Array): string {
  return Array.from(data).map(b => b.toString(16).padStart(2, '0')).join('');
}

// Helper to load golden file
function loadGolden(name: string): Uint8Array | null {
  const filePath = path.join(GOLDEN_DIR, `${name}.bin`);
  try {
    return new Uint8Array(fs.readFileSync(filePath));
  } catch {
    return null;
  }
}

describe('TypeScript Interoperability Tests', () => {
  describe('NestedMessage', () => {
    it('encodes and decodes correctly', () => {
      const writer = new Writer();
      encodeNestedMessage(writer, TestData.nestedMessage);
      const encoded = writer.bytes();

      console.log('NestedMessage encoded:', toHex(encoded));

      const reader = new Reader(encoded);
      const decoded = decodeNestedMessage(reader);

      expect(decoded.name).toBe(TestData.nestedMessage.name);
      expect(decoded.value).toBe(TestData.nestedMessage.value);
    });

    it('matches golden file', () => {
      const golden = loadGolden('nested_message');
      if (!golden) {
        console.log('Golden file not found, skipping');
        return;
      }

      // Decode golden file
      const reader = new Reader(golden);
      const decoded = decodeNestedMessage(reader);

      expect(decoded.name).toBe(TestData.nestedMessage.name);
      expect(decoded.value).toBe(TestData.nestedMessage.value);
    });
  });

  describe('ScalarTypes', () => {
    it('encodes and decodes correctly', () => {
      const writer = new Writer();
      encodeScalarTypes(writer, TestData.scalarTypes);
      const encoded = writer.bytes();

      console.log('ScalarTypes encoded:', toHex(encoded));
      console.log('ScalarTypes size:', encoded.length, 'bytes');

      const reader = new Reader(encoded);
      const decoded = decodeScalarTypes(reader);

      expect(decoded.boolVal).toBe(TestData.scalarTypes.boolVal);
      expect(decoded.int32Val).toBe(TestData.scalarTypes.int32Val);
      expect(decoded.int64Val).toBe(TestData.scalarTypes.int64Val);
      expect(decoded.uint32Val).toBe(TestData.scalarTypes.uint32Val);
      expect(decoded.uint64Val).toBe(TestData.scalarTypes.uint64Val);
      expect(decoded.float32Val).toBeCloseTo(TestData.scalarTypes.float32Val, 4);
      expect(decoded.float64Val).toBe(TestData.scalarTypes.float64Val);
      expect(decoded.stringVal).toBe(TestData.scalarTypes.stringVal);
      expect(toHex(decoded.bytesVal)).toBe(toHex(TestData.scalarTypes.bytesVal));
    });

    it('decodes golden file correctly', () => {
      const golden = loadGolden('scalar_types');
      if (!golden) {
        console.log('Golden file not found, skipping');
        return;
      }

      console.log('Golden ScalarTypes hex:', toHex(golden));

      const reader = new Reader(golden);
      const decoded = decodeScalarTypes(reader);

      expect(decoded.boolVal).toBe(TestData.scalarTypes.boolVal);
      expect(decoded.int32Val).toBe(TestData.scalarTypes.int32Val);
      expect(decoded.int64Val).toBe(TestData.scalarTypes.int64Val);
      expect(decoded.uint32Val).toBe(TestData.scalarTypes.uint32Val);
      expect(decoded.uint64Val).toBe(TestData.scalarTypes.uint64Val);
      expect(decoded.float32Val).toBeCloseTo(TestData.scalarTypes.float32Val, 4);
      expect(decoded.float64Val).toBe(TestData.scalarTypes.float64Val);
      expect(decoded.stringVal).toBe(TestData.scalarTypes.stringVal);
      expect(toHex(decoded.bytesVal)).toBe(toHex(TestData.scalarTypes.bytesVal));
    });
  });

  describe('AllFieldNumbers', () => {
    it('encodes and decodes correctly', () => {
      const writer = new Writer();
      encodeAllFieldNumbers(writer, TestData.allFieldNumbers);
      const encoded = writer.bytes();

      console.log('AllFieldNumbers encoded:', toHex(encoded));

      // Decode - V2 format uses end marker
      const reader = new Reader(encoded);
      const decoded: Partial<AllFieldNumbers> = {};

      while (reader.hasMore) {
        const tag = reader.readTag();
        if (tag.fieldNumber === 0) break; // End marker

        switch (tag.fieldNumber) {
          case 1: decoded.field1 = reader.readSVarint(); break;
          case 15: decoded.field15 = reader.readSVarint(); break;
          case 16: decoded.field16 = reader.readSVarint(); break;
          case 127: decoded.field127 = reader.readSVarint(); break;
          case 128: decoded.field128 = reader.readSVarint(); break;
          case 1000: decoded.field1000 = reader.readSVarint(); break;
          default: reader.skipField(tag.wireType);
        }
      }

      expect(decoded.field1).toBe(TestData.allFieldNumbers.field1);
      expect(decoded.field15).toBe(TestData.allFieldNumbers.field15);
      expect(decoded.field16).toBe(TestData.allFieldNumbers.field16);
      expect(decoded.field127).toBe(TestData.allFieldNumbers.field127);
      expect(decoded.field128).toBe(TestData.allFieldNumbers.field128);
      expect(decoded.field1000).toBe(TestData.allFieldNumbers.field1000);
    });

    it('decodes golden file correctly', () => {
      const golden = loadGolden('all_field_numbers');
      if (!golden) {
        console.log('Golden file not found, skipping');
        return;
      }

      console.log('Golden AllFieldNumbers hex:', toHex(golden));

      // Decode - V2 format uses end marker
      const reader = new Reader(golden);
      const decoded: Partial<AllFieldNumbers> = {};

      while (reader.hasMore) {
        const tag = reader.readTag();
        if (tag.fieldNumber === 0) break; // End marker

        switch (tag.fieldNumber) {
          case 1: decoded.field1 = reader.readSVarint(); break;
          case 15: decoded.field15 = reader.readSVarint(); break;
          case 16: decoded.field16 = reader.readSVarint(); break;
          case 127: decoded.field127 = reader.readSVarint(); break;
          case 128: decoded.field128 = reader.readSVarint(); break;
          case 1000: decoded.field1000 = reader.readSVarint(); break;
          default: reader.skipField(tag.wireType);
        }
      }

      expect(decoded.field1).toBe(TestData.allFieldNumbers.field1);
      expect(decoded.field15).toBe(TestData.allFieldNumbers.field15);
      expect(decoded.field16).toBe(TestData.allFieldNumbers.field16);
      expect(decoded.field127).toBe(TestData.allFieldNumbers.field127);
      expect(decoded.field128).toBe(TestData.allFieldNumbers.field128);
      expect(decoded.field1000).toBe(TestData.allFieldNumbers.field1000);
    });
  });

  describe('Wire Format Primitives', () => {
    it('varint encoding matches Go', () => {
      const testCases = [
        { value: 0, expected: '00' },
        { value: 1, expected: '01' },
        { value: 127, expected: '7f' },
        { value: 128, expected: '8001' },
        { value: 300, expected: 'ac02' },
        { value: 16384, expected: '808001' },
      ];

      for (const tc of testCases) {
        const writer = new Writer();
        writer.writeVarint(tc.value);
        const hex = toHex(writer.bytes());
        expect(hex).toBe(tc.expected);
      }
    });

    it('zigzag encoding matches Go', () => {
      const testCases = [
        { value: 0, expected: '00' },
        { value: -1, expected: '01' },
        { value: 1, expected: '02' },
        { value: -2, expected: '03' },
        { value: 2, expected: '04' },
        { value: -42, expected: '53' },
        { value: 42, expected: '54' },
      ];

      for (const tc of testCases) {
        const writer = new Writer();
        writer.writeSVarint(tc.value);
        const hex = toHex(writer.bytes());
        expect(hex).toBe(tc.expected);
      }
    });
  });
});
